model-optimizer/extensions/middle/TFResizeBilinearToInterpolateV4.py

```diff
@@ -0,0 +1,123 @@
+ limitations under the License.
+"""
+
+import logging as log
+
+import numpy as np
+
+from extensions.ops.interpolate import Interpolate
+from mo.front.common.layout import get_height_dim, get_width_dim
+from mo.front.common.partial_infer.utils import int64_array
+from mo.front.tf.graph_utils import create_op_with_const_inputs
+from mo.graph.graph import Graph, Node, rename_nodes
+from mo.middle.replacement import MiddleReplacementPattern
+
+
+def replace_resize_bilinear(graph: Graph, resize: Node):
+    resize_name = resize.soft_get('name', resize.id)
+    log.debug("Converting of TF ResizeBilinear to Interpolate-4 is triggered for node {}.".format(resize_name))
+
+    input_shape = resize.in_port(0).data.get_shape()
+    input_rank = len(input_shape)
+    if input_rank != 4:
+        log.warning('The input shape is not 4D for op with name {}'.format(resize_name))
+        return
+
+    num_of_inputs = len([port for port in resize.in_ports().values() if not port.disconnected()])
+    assert num_of_inputs == 2, \
+        "Number of inputs of TFResizeBilinear (with name {}) should be equal to 2".format(resize_name)
+
+    new_sizes_value = resize.in_port(1).data.get_value()
+    assert new_sizes_value is not None, "Node {} with op {} has no value in input port 1".format(resize_name, resize.op)
+
+    attrs_msg = "If half_pixel_centers attribute of the node {} with op {} is True, " \
+                "the attribute align_corners must be False"
+    assert not resize.half_pixel_centers or (resize.half_pixel_centers and not resize.align_corners), \
+        attrs_msg.format(resize_name, resize.op)
+
+    out_height = new_sizes_value[0]
+    out_width = new_sizes_value[1]
+
+    layout = graph.graph['layout']
+    height_dim = get_height_dim(layout, input_rank)
+    width_dim = get_width_dim(layout, input_rank)
+
+    input_height = input_shape[height_dim]
+    input_width = input_shape[width_dim]
+
+    align_corners = resize.align_corners
+    half_pixel_centers = resize.half_pixel_centers
+
+    if align_corners:
+        coordinate_transformation_mode = 'align_corners'
+    elif half_pixel_centers:
+        coordinate_transformation_mode = 'half_pixel'
+    else:
+        coordinate_transformation_mode = 'asymmetric'
+
+    shape_calculation_mode = 'sizes'
+    scales_data = np.array([out_height / input_height, out_width / input_width], dtype=np.float32)
+
+    if out_height > 1 and out_width <= 1 and align_corners:
+        shape_calculation_mode = 'scales'
+        coordinate_transformation_mode = 'asymmetric'
+        scales_data = np.array([(out_height - 1) / (input_height - 1), out_width / input_width], dtype=np.float32)
+    elif out_height <= 1 and out_width > 1 and align_corners:
+        shape_calculation_mode = 'scales'
+        coordinate_transformation_mode = 'asymmetric'
+        scales_data = np.array([out_height / input_height, (out_width - 1) / (input_width - 1)], dtype=np.float32)
```

Comment:
This is not a robust solution. If the input for the operation is changed in the runtime then these logic is not valid anymore. Let's discuss this during the meeting.

Also, ideally the operation should be replaced during the front phase.

Reply:
The most robust solution is to introduce a new coordinate transformation mode into the operation `Interpolate`, `tf_align_corners`, to take into an account the specific of the TF attributes `align_corners`. The meaning of the attribute `align_corners` of ONNX Resize and the meaning of the attribute `align_corners` of TF `ResizeBilinear` and `ResizeNearestNeighbor` are different.

Reply:
The transformation was moved into the front stage.